midp/location.js

```diff
@@ -0,0 +1,140 @@
+    return J2ME.newString(Location.PROVIDER_NAME);
+};
+
+addUnimplementedNative("com/sun/j2me/location/CriteriaImpl.initNativeClass.()V");
+
+Native["com/sun/j2me/location/PlatformLocationProvider.getBestProviderByCriteriaImpl.(Lcom/sun/j2me/location/CriteriaImpl;)Z"] = function(criteria) {
+    criteria.providerName = J2ME.newString(Location.PROVIDER_NAME);
+    return 1;
+};
+
+addUnimplementedNative("com/sun/j2me/location/LocationProviderInfo.initNativeClass.()V");
+addUnimplementedNative("com/sun/j2me/location/LocationInfo.initNativeClass.()V");
+
+Native["com/sun/j2me/location/PlatformLocationProvider.open.(Ljava/lang/String;)I"] = function(name) {
+    var provider = new LocationProvider();
+    provider.start();
+    var id = Location.Providers.nextId;
+    Location.Providers.nextId = Location.Providers.nextId % 0xff + 1;
+    Location.Providers[id] = provider;
+    return id;
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.resetImpl.(I)V"] = function(providerId) {
+    var provider = Location.Providers[providerId];
+    provider.stop();
+    Location.Providers[providerId] = null;
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.getCriteria.(Ljava/lang/String;Lcom/sun/j2me/location/LocationProviderInfo;)Z"] = function(name, criteria) {
+    criteria.canReportAltitude = 1;
+    criteria.canReportSpeedCource = 1;
+    criteria.averageResponseTime = 10000;
+    return 1;
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.setUpdateIntervalImpl.(II)V"] = function(providerId, interval) {
+    console.warn("com/sun/j2me/location/PlatformLocationProvider.setUpdateIntervalImpl.(II)V not implemented");
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.getLastLocationImpl.(ILcom/sun/j2me/location/LocationInfo;)Z"] = function(providerId, locationInfo) {
+    var provider = Location.Providers[providerId];
+    var pos = provider.position;
+    locationInfo.isValid = 1;
+    locationInfo.timestamp = Long.fromNumber(pos.timestamp);
+    locationInfo.latitude = pos.latitude;
+    locationInfo.longitude = pos.longitude;
+    locationInfo.altitude = Math.fround(pos.altitude);
+    locationInfo.horizontalAccuracy =  Math.fround(pos.horizontalAccuracy);
+    locationInfo.verticalAccuracy =  Math.fround(pos.verticalAccuracy);
+    locationInfo.speed = Math.fround(pos.speed);
+    locationInfo.course =  Math.fround(pos.course);
+    locationInfo.method = 0;
+    return 1;
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.getStateImpl.(I)I"] = function(providerId) {
+    var provider = Location.Providers[providerId];
+    return provider.state;
+};
+
+Native["com/sun/j2me/location/PlatformLocationProvider.waitForNewLocation.(IJ)Z"] = function(providerId, timeout) {
+    asyncImpl("Z", new Promise(function(resolve, reject) {
+        var provider = Location.Providers[providerId];
+        provider.requestData().then(resolve.bind(null, 1));
+        setTimeout(resolve.bind(null, 0), timeout);
```

Comment:
It shouldn't be necessary to bind the *resolve* function, although it doesn't cause any harm.